test/distributed/test_ddp_with_rpc.py

```diff
@@ -0,0 +1,541 @@
+    )
+    console.setFormatter(formatter)
+    console.setLevel(logging.DEBUG if "debug" in os.environ else logging.INFO)
+    # add the handlers to the logger
+    logger.addHandler(console)
+    logger.propagate = False
+    return logger
+
+
+gLogger = init_logger()
+
+
+class FeatureSet(NamedTuple):
+    """ A feature set has 2 types of features"""
+
+    dense_features: torch.Tensor
+    sparse_features: torch.LongTensor
+    labels: torch.LongTensor
+
+
+def _call_method(method, rref, *args, **kwargs):
+    return method(rref.local_value(), *args, **kwargs)
+
+
+def _remote_method(method, rref, *args, **kwargs):
+    args_tup = tuple([method, rref] + list(args))
+    return rpc.rpc_sync(
+        rref.owner(), _call_method, args=args_tup, kwargs=kwargs
+    )
+
+
+def _remote_method_async(method, rref, *args, **kwargs):
+    args_tup = tuple([method, rref] + list(args))
+    return rpc.rpc_async(
+        rref.owner(), _call_method, args=args_tup, kwargs=kwargs
+    )
+
+
+def check_process_group_type(
+    process_name: str, process_group: dist.ProcessGroup
+):
+    if not isinstance(process_group, dist.ProcessGroup):
+        gLogger.warning(
+            f"The process group in {process_name} process is of type {type(process_group)}"
+        )
+
+
+class RpcContext:
+    def __init__(
+        self, name: str, rank: int, world_size: int, group_name: str = ""
+    ):
+        self.name = name
+        self.rank = rank
+        self.world_size = world_size
+        self.group_name = group_name
+
+    def __enter__(self):
+        gLogger.info(
+            f"Initing RPC [{self.group_name}] by {self.name} with rank "
+            f"#{self.rank} out of {self.world_size} peers."
+        )
+        rpc.init_rpc(name=self.name, rank=self.rank, world_size=self.world_size)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        gLogger.info(
+            f"Shutting down RPC group [{self.group_name}] from process {self.name}."
+        )
+        if exc_type is not None:
+            raise exc_value
+        rpc.shutdown()
+
+
+class ProcessGroupContext:
+    def __init__(self, name: str, ranks: [int], group_name: str = ""):
+        self.name = name
+        self.ranks = ranks
+        self.group_name = group_name
+        self.group = None
+
+    def process_group(self) -> dist.ProcessGroup:
+        return self.group
+
+    def __enter__(self) -> dist.ProcessGroup:
+        gLogger.info(
+            f"Initing process group [{self.group_name}] by {self.name} with ranks {self.ranks}"
+        )
+        self.group = dist_c10d.new_group(ranks=self.ranks)
+        return self.group
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        gLogger.info(
+            f"Destroy process group [{self.group_name}] from process {self.name}."
+        )
+        if exc_type is not None:
+            raise exc_value
+        dist.destroy_process_group(self.group)
```

Comment:
Can add destroy_process_group at the end of `dist_init` if the process group was initialized and then we don't need this additional context.

Reply:
Will do.